toolchain/base/kind_switch.h

```diff
@@ -50,65 +50,116 @@ template <typename... Ts>
-  struct StdVariantEnumValue<std::variant<__VA_ARGS__>> {                     \
-    enum EnumType {                                                           \
-      CARBON_FOR_EACH(CARBON_INTERNAL_MAKE_ENUM_VALUE, CARBON_FOR_EACH_COMMA, \
-                      __VA_ARGS__)                                            \
-    };                                                                        \
-    using Type = EnumType;                                                    \
-  }
-
-CARBON_INTERNAL_MAKE_ENUM(T0);
-CARBON_INTERNAL_MAKE_ENUM(T0, T1);
-CARBON_INTERNAL_MAKE_ENUM(T0, T1, T2);
-CARBON_INTERNAL_MAKE_ENUM(T0, T1, T2, T3);
-CARBON_INTERNAL_MAKE_ENUM(T0, T1, T2, T3, T4);
-CARBON_INTERNAL_MAKE_ENUM(T0, T1, T2, T3, T4, T5);
-CARBON_INTERNAL_MAKE_ENUM(T0, T1, T2, T3, T4, T5, T6);
-CARBON_INTERNAL_MAKE_ENUM(T0, T1, T2, T3, T4, T5, T6, T7);
-CARBON_INTERNAL_MAKE_ENUM(T0, T1, T2, T3, T4, T5, T6, T7, T8);
-CARBON_INTERNAL_MAKE_ENUM(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9);
-CARBON_INTERNAL_MAKE_ENUM(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10);
-CARBON_INTERNAL_MAKE_ENUM(T0, T1, T2, T3, T4, T5, T6, T7, T8, T9, T10, T11);
-
-#undef CARBON_INTERNAL_MAKE_ENUM_VALUE
-#undef CARBON_INTERNAL_MAKE_TYPENAME
-#undef CARBON_INTERNAL_MAKE_ENUM
-
-template <typename T>
-using StdVariantEnum = StdVariantEnumValue<std::decay_t<T>>::Type;
-
-// Given `CARBON_KIND_SWITCH(value)` this returns `value.kind()` to switch on.
+// Generate StdVariantTypeMap specializations for each number of types in the
+// std::variant<...> type list.
+CARBON_INTERNAL_TYPE_MAP(0);
+CARBON_INTERNAL_TYPE_MAP(0, 1);
+CARBON_INTERNAL_TYPE_MAP(0, 1, 2);
+CARBON_INTERNAL_TYPE_MAP(0, 1, 2, 3);
+CARBON_INTERNAL_TYPE_MAP(0, 1, 2, 3, 4);
+CARBON_INTERNAL_TYPE_MAP(0, 1, 2, 3, 4, 5);
+CARBON_INTERNAL_TYPE_MAP(0, 1, 2, 3, 4, 5, 6);
+CARBON_INTERNAL_TYPE_MAP(0, 1, 2, 3, 4, 5, 6, 7);
+CARBON_INTERNAL_TYPE_MAP(0, 1, 2, 3, 4, 5, 6, 7, 8);
+CARBON_INTERNAL_TYPE_MAP(0, 1, 2, 3, 4, 5, 6, 7, 8, 9);
+CARBON_INTERNAL_TYPE_MAP(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
+CARBON_INTERNAL_TYPE_MAP(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11);
+
+#undef CARBON_INTERNAL_IDENTIFIER
+#undef CARBON_INTERNAL_IDENTIFIERS
+#undef CARBON_INTERNAL_TYPENAME
+#undef CARBON_INTERNAL_TYPENAMES
+#undef CARBON_INTERNAL_ENUM_NAME
+#undef CARBON_INTERNAL_TYPES_TO_ENUM_NAMES
+#undef CARBON_INTERNAL_TYPE_TO_ENUM_NAME
+#undef CARBON_INTERNAL_TYPE_MAP
+
+// Uses the above `CARBON_INTERNAL_TYPE_MAP` expansions to make an enum with a
+// value for each type in a std::variant<...> type list.
+template <typename StdVariant>
+using StdVariantEnum = StdVariantTypeMap<std::decay_t<StdVariant>>::EnumType;
+
+// Uses the `CARBON_INTERNAL_TYPE_MAP` expanstions to find the enum value in
+// `StdVariantEnum` for a given type `T` in the type list of a
+// std::variant<...>.
+template <typename T, typename StdVariant>
+[[maybe_unused]] static constexpr auto CaseValueOfTypeInStdVariant =
```

Comment:
Why `static`? This creates an opportunity for odr issues. Also, given that this is only used once, maybe we could save a variable template instantiation and inline it?

Reply:
Oh.. right. Same with the ones earlier in the file. static removed from them all.

I would prefer to keep it here though, as it is the "public api" of the EnumValue construct, so it can live near it.

Reply:
OK, works for me.